cmd/ceremony/main.go

```diff
@@ -0,0 +1,280 @@
+		}
+		if err := verifyProfile(cc.CertificateProfile, true); err != nil {
+			return fmt.Errorf("invalid certificate-profile: %s", err)
+		}
+	case "intermediate":
+		if cc.KeyLabel == "" {
+			return errors.New("key-label is required")
+		}
+		if cc.KeyID == "" {
+			return errors.New("key-id is required")
+		}
+		if cc.KeyType != "" {
+			return errors.New("key-type is not used for intermediate ceremonies")
+		}
+		if cc.ECDSACurve != "" {
+			return errors.New("ecdsa-curve is not used for intermediate ceremonies")
+		}
+		if cc.PublicKeyPath == "" {
+			return errors.New("public-key-path is required")
+		}
+		if cc.CertificatePath == "" {
+			return errors.New("certificate-path is required")
+		}
+		if cc.IssuerPath == "" {
+			return errors.New("issuer-path is required")
+		}
+		if cc.CertificateProfile == nil {
+			return errors.New("certificate-profile is required")
+		}
+		if err := verifyProfile(cc.CertificateProfile, true); err != nil {
+			return fmt.Errorf("invalid certificate-profile: %s", err)
+		}
+	case "key":
+		if cc.KeyLabel == "" {
+			return errors.New("key-label is required")
+		}
+		if cc.KeyID != "" {
+			return errors.New("key-id is not used for key ceremonies")
+		}
+		if cc.KeyType == "" {
+			return errors.New("key-type is required")
+		}
+		if cc.KeyType != "rsa" && cc.KeyType != "ecdsa" {
+			return errors.New("key-type can only be 'rsa' or 'ecdsa'")
+		}
+		if cc.KeyType == "rsa" && cc.ECDSACurve != "" {
+			return errors.New("if key-type = \"rsa\" then ecdsa-curve is not used")
+		}
+		if cc.KeyType == "ecdsa" && cc.ECDSACurve == "" {
+			return errors.New("if key-type = \"ecdsa\" then ecdsa-curve is required")
+		}
+		if cc.PublicKeyPath == "" {
+			return errors.New("public-key-path is required")
+		}
+		if cc.IssuerPath != "" {
+			return errors.New("issuer-path is not used for key ceremonies")
+		}
+		if cc.CertificatePath != "" {
+			return errors.New("certificate-path is not used for key ceremonies")
+		}
+		if cc.CertificateProfile != nil {
+			return errors.New("certificate-profile is not used for key ceremonies")
+		}
+	}
+
+	return nil
+}
+
+func main() {
+	configPath := flag.String("config", "", "Path to ceremony configuration file")
+	flag.Parse()
+
+	if *configPath == "" {
+		log.Fatal("--config is required")
+	}
+	configBytes, err := ioutil.ReadFile(*configPath)
+	if err != nil {
+		log.Fatalf("Failed to read config file: %s", err)
+	}
+	var config ceremonyConfig
+	err = json.Unmarshal(configBytes, &config)
+	if err != nil {
+		log.Fatalf("Failed to parse config: %s", err)
+	}
+
+	if err = config.Validate(); err != nil {
+		log.Fatalf("Failed to validate config: %s", err)
+	}
+
+	ctx, session, err := pkcs11helpers.Initialize(config.PKCS11Module, config.KeySlot, "")
+	if err != nil {
+		log.Fatalf("Failed to setup session and PKCS#11 context for slot %d: %s", config.KeySlot, err)
+	}
+	log.Printf("Opened PKCS#11 session for slot %d\n", config.KeySlot)
+
+	switch config.CeremonyType {
```

Comment:
I think the `case` bodies in this `switch` are pretty large and could probably be pulled out into separate functions. That would probably make it easier to test too since `main` is a little bit cumbersome sometimes.